ARROW-12392: [C++] Restore asynchronous streaming CSV reader

westonpace opened the pull request:
This restores ARROW-11887.  The only difference is that I kept the async path out of RecordBatchReader.  I now know the file formats do not use RecordBatchReader for the read so there was no need to add it in there yet.

It also now properly consumes (using RunInSerialExecutor) the reader via the nested parallel paths in the synchronous scanner.

westonpace commented:
Keeping in draft until #10064 is resolved or else it'll just reintroduce the R deadlock issue.

lidavidm commented on cpp/src/arrow/util/async_generator.h:
```diff
@@ -1475,11 +1475,44 @@ static Result<AsyncGenerator<T>> MakeBackgroundGenerator(
+          typename Generator = internal::call_traits::return_type<GeneratorFactory>,
+          typename FT = typename Generator::result_type,
+          typename T = typename FT::ValueType>
+Result<Iterator<T>> MakeGeneratorIterator(GeneratorFactory factory) {
```
It would be nice to explain the purpose of this class and whether it's reentrant or not (well, the question is moot presumably because it's blocking). This essentially 1) lets you parameterize a generator on an executor (which is guaranteed to resolve to a SerialExecutor) and then 2) synchronously blocks on the generator via the SerialExecutor, effectively running all work on the main thread?

lidavidm commented on cpp/src/arrow/util/async_generator_test.cc:
```diff
@@ -1102,6 +1105,46 @@ TEST(TestAsyncUtil, ReadaheadFailed) {
+        auto back = MakeTransferredGenerator(std::move(to_io), executor);
+        return back;
+      }));
+  ASSERT_OK_AND_ASSIGN(auto actual, it.ToVector());
```
Just to check my understanding, one call here (when IsSlow()) will synchronously sleep on the CPU executor, attempt to transfer to the I/O executor (which is a no-op as the future is already complete), then transfer back to the CPU executor (also a no-op)?

westonpace replied:
N/A, code removed.

lidavidm commented on cpp/src/arrow/dataset/scanner.h:
```diff
@@ -152,6 +152,9 @@ class ARROW_DS_EXPORT ScanTask {
   /// resulting from the Scan. Execution semantics are encapsulated in the
   /// particular ScanTask implementation
   virtual Result<RecordBatchIterator> Execute() = 0;
```
So to make sure, SafeExecute/SafeVisit are just a way to safely integrate async readers while we still have SyncScanner?

lidavidm replied:
(Note #10062 removes the ScanTask bindings from Python, leaving just CGlib.)

westonpace replied:
Correct.  These methods can go away when `SyncScanner` goes away.  Removing the `ScanTask` bindings won't be enough (although I'm happy to hear they are removed) because `SyncScanner::ToTable` and the synchronous dataset write both rely on scanning being safe to call (ala nested deadlock) from a CPU executor thread.

westonpace replied:
Also, this will probably only be needed for CSV.  With parquet & IPC you created the async version in parallel with the sync verison.  With CSV there is just one (async) code path.

lidavidm commented on cpp/src/arrow/dataset/file_csv.cc:
```diff
@@ -154,9 +180,26 @@ class CsvScanTask : public ScanTask {
+
+  Future<RecordBatchVector> SafeExecute(internal::Executor* executor) override {
+    auto reader_fut =
+        OpenReaderAsync(source_, *format_, options(), executor, options()->pool);
```
So `executor` here will be a SerialExecutor. Isn't that a behavior change? Previously, `StreamingReader::Make` would use the CPU thread pool. Now we're having it use the main thread.

westonpace replied:
`SerialExecutor` assumes the current thread which may be from the CPU thread pool.  So, for example, `SyncScanner::ToTable` will continue to run `SafeExecute` on the CPU thread pool.

lidavidm replied:
Ahh, I see now - before, while a cpu_executor was passed around in StreamingReader, it wasn't actually used. (MakeStreamingReader accepts an executor, but never uses it.) Now it is used, but it'll be a SerialExecutor, so we remain single-threaded as desired.

lidavidm reviewed (commented):
Just one nit that I missed before (sorry).

westonpace commented on cpp/src/arrow/dataset/file_base.cc:
```diff
@@ -480,20 +479,14 @@ Status FileSystemDataset::Write(const FileSystemDatasetWriteOptions& write_optio
   }
   return task_group->Finish();
-}
+}  // namespace dataset
```
Fixed.  It's frustrating that clang-format is smart enough to add these but not smart enough to remove them.

westonpace commented:
Rebased and addressed remaining PR nit.  Perhaps best to wait until after the release has been finalized before merging this as it does affect the sync path.

westonpace commented:
Rebase.  Appveyor failure appears unrelated.  Now that 4.0 is past the RC phase I'd appreciate merge or re-review.

lidavidm reviewed (approved):
Originally I wanted to wait for #10166 since I was worried there would be a lot of conflicts. However it looks like #10166 needs some more work + the conflicts are limited to scanner_internal.h so maybe we should just merge this. CC @bkietz as a heads up.
